README.md
Location Share
==============

A simple Android application that shares your current location.

<a href="https://play.google.com/store/apps/details?id=ca.cmetcalfe.locationshare">
  <img src="https://play.google.com/intl/en_us/badges/images/generic/en_badge_web_generic.png"
       alt="Get it on Google Play"
       height="80"/>
</a>
<a href="https://f-droid.org/packages/ca.cmetcalfe.locationshare/">
  <img src="https://fdroid.gitlab.io/artwork/badge/get-it-on.png"
       alt="Get it on F-Droid"
       height="80">
</a>

Location Share doesn't collect any data from you, doesn't require internet access, doesn't require
setting up an account, doesn't track your location in the background, doesn't auto-share it, and
doesn't send it to anyone except those you explicitly tell it to.

It can share your location using any application that supports sharing plain text. This includes
(but isn't limited to) sharing via SMS, email, social networks, note taking applications, etc.

Simply launch the application and wait until the device homes in on your location. You'll then have
the option to view the location on a map, share it, or copy it to the clipboard.

------------------------------------

Thanks to [all the contributors](CONTRIBUTORS.md) who've helped improve this project!


CHANGELOG.md
# Location Share Changelog

### [v1.4.2] - 2022-03-25
- Updates Norwegian translations - thanks [FTno](https://github.com/FTno)
- Add Brazilian Portuguese translation
- App maintenance, library updates, and optimizations - thanks [@TacoTheDank](https://github.com/TacoTheDank)
- Target Android S (API 31)

### [v1.4.1] - 2021-05-18
- Adds Russian translation - thanks [@ashed](https://github.com/ashed)

### [v1.4.0] - 2021-04-24
- Adds a DMS representation of coordinates
- Adds the ability to share coordinates in a raw decimal form "{lat}, {lon}"
- Adds Czech translation - thanks [@thenhustey](https://github.com/thenhustey)
- Adds Indonesian translation - thanks [@zmni](https://github.com/zmni)
- Fixes an issue that caused stale locations to be shown
- Target Android R (API 30)

### [v1.3.2] - 2019-01-14
- Target Android Pie (API 28)

### [v1.3.1] - 2018-09-14
- Added 2 missing translated strings

### [v1.3] - 2018-09-09
- Added the ability to configure the URL type in a settings menu - thanks [@mahula](https://github.com/mahula)
- Added translations for Bulgarian, Danish, Dutch, French, Japanese, Korean, Norwegian, and Swedish (see CONTRIBUTORS.md)
- Updated the supported SDK range (now 14 to 27, was 9 to 25)

### [v1.2] - 2017-02-06
- Fix lat/long display - thanks [@mahula](https://github.com/mahula)
- Change the precision of lat/long to be more realistic

### [v1.1] - 2017-02-04
- Add a choice of 3 URL types: Google Maps, OpenStreetMap, GeoURI
- Add German translations - thanks [@mahula](https://github.com/mahula)

### [v1.0] - 2016-11-03
- Initial version

 [v1.4.2]: https://github.com/pR0Ps/LocationShare/compare/v1.4.1...v1.4.2
 [v1.4.1]: https://github.com/pR0Ps/LocationShare/compare/v1.4.0...v1.4.1
 [v1.4.0]: https://github.com/pR0Ps/LocationShare/compare/v1.3.2...v1.4.0
 [v1.3.2]: https://github.com/pR0Ps/LocationShare/compare/v1.3.1...v1.3.2
 [v1.3.1]: https://github.com/pR0Ps/LocationShare/compare/v1.3...v1.3.1
 [v1.3]: https://github.com/pR0Ps/LocationShare/compare/v1.2...v1.3
 [v1.2]: https://github.com/pR0Ps/LocationShare/compare/v1.1...v1.2
 [v1.1]: https://github.com/pR0Ps/LocationShare/compare/v1.0...v1.1
 [v1.0]: https://github.com/pR0Ps/LocationShare/releases/tag/v1.0


src/com/pr0ps/locationshare/LocationShare.java
package com.pr0ps.locationshare;

import java.text.DecimalFormat;
import java.text.MessageFormat;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.view.Menu;
import android.view.View;
import android.widget.TextView;

public class LocationShare extends Activity {

    LocationManager mlocManager;
    LocationListener mlocListener;
    Location location;
    long lastFix;

    private Handler handler = new Handler();
    private Runnable updateUITask = new Runnable() {
        public void run() {
            updateDisplay();
            handler.postDelayed(this, 100);
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_location_share);

        mlocManager = (LocationManager)getSystemService(Context.LOCATION_SERVICE);
        mlocListener = new MyLocationListener(this);
        location = mlocManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
        if (location != null){
            lastFix = location.getTime();
        }

        handler.post(updateUITask);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.location_share, menu);
        return true;
    }

    @Override
    protected void onStop(){
        super.onStop();
        mlocManager.removeUpdates(mlocListener);
    }

    @Override
    protected void onResume(){
        super.onResume();
        mlocManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, mlocListener);
        mlocManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0, mlocListener);
    }

    public void shareLocation (View view){
        if (this.location == null){
            return;
        }

        String link = formatLocation("http://maps.google.com/?q={0},{1}", this.location);

        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_SEND);
        intent.putExtra(Intent.EXTRA_TEXT, link);
        intent.setType("text/plain");
        startActivity(Intent.createChooser(intent, getResources().getText(R.string.send_via)));
    }

    public void viewLocation (View view){
        if (this.location == null){
            return;
        }

        String uri = formatLocation("geo:{0},{1}?q={0},{1}", this.location);

        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
        startActivity(Intent.createChooser(intent, getResources().getText(R.string.view_via)));
    }

    public void updateLocation(Location location){
        //TODO: Check if new location is 'better'
        this.location = location;
        this.lastFix = location.getTime();
        updateDisplay();
    }

    public void updateDisplay(){

        if (this.location != null){
            long timeElapsed = System.currentTimeMillis() - this.lastFix;
            ((TextView)findViewById(R.id.timeText)).setText(stringifyTime(timeElapsed/1000));

            ((TextView)findViewById(R.id.latitudeText)).setText("" + this.location.getLatitude());
            ((TextView)findViewById(R.id.longitudeText)).setText("" + this.location.getLongitude());
            ((TextView)findViewById(R.id.accuracyText)).setText(new DecimalFormat("###.00").format(this.location.getAccuracy()) + "m");
        }
        else{
            ((TextView)findViewById(R.id.latitudeText)).setText("N/A");
            ((TextView)findViewById(R.id.longitudeText)).setText("N/A");
            ((TextView)findViewById(R.id.accuracyText)).setText("N/A");
            ((TextView)findViewById(R.id.timeText)).setText("N/A");
        }

    }

    private String formatLocation(String s, Location l){
        //Hack to get around MessageFormat precision weirdness
        return MessageFormat.format(s, ""+l.getLatitude(), ""+l.getLongitude(), ""+l.hasAccuracy());
    }

    public static String stringifyTime(long seconds) {

        //int days = (int)(seconds/86400);
        //seconds -= days * 86400;
        int hours = (int)(seconds/3600);
        seconds -= hours * 3600;
        int minutes = (int)(seconds/60);
        seconds -= minutes * 60;

        //Fix negative values (rounding)
        seconds = Math.max(0, seconds);

        return MessageFormat.format("{0,number,00}:{1,number,00}:{2,number,00}", hours, minutes, seconds);
    }

}


app/src/main/java/ca/cmetcalfe/locationshare/SettingsActivity.java
package ca.cmetcalfe.locationshare;

import android.os.Bundle;
import android.view.MenuItem;

import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.PreferenceFragmentCompat;

import ca.cmetcalfe.locationshare.databinding.ActivitySettingsBinding;

public class SettingsActivity extends AppCompatActivity {

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final ActivitySettingsBinding binding =
                ActivitySettingsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.settingsToolbar.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        getSupportFragmentManager().beginTransaction()
                .replace(binding.settingsContainer.getId(), new SettingsFragment())
                .commit();
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public static class SettingsFragment extends PreferenceFragmentCompat {
        @Override
        public void onCreatePreferences(final Bundle savedInstanceState, final String rootKey) {
            addPreferencesFromResource(R.xml.preferences);
        }
    }
}


app/src/main/java/ca/cmetcalfe/locationshare/MainActivity.java
package ca.cmetcalfe.locationshare;

import android.Manifest;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.SystemClock;
import android.provider.Settings;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import java.text.MessageFormat;
import java.util.Locale;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts.RequestPermission;
import androidx.appcompat.app.AlertDialog.Builder;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.preference.PreferenceManager;

import ca.cmetcalfe.locationshare.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private LocationManager locManager;
    private Location lastLocation;
    private ActivityMainBinding binding;
    private SharedPreferences sharedPrefs;

    private final LocationListener locListener = new LocationListener() {
        public void onLocationChanged(Location loc) {
            updateLocation(loc);
        }

        public void onProviderEnabled(String provider) {
            updateLocation();
        }

        public void onProviderDisabled(String provider) {
            updateLocation();
        }
    };

    // ----------------------------------------------------
    // Android Lifecycle
    // ----------------------------------------------------
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.toolbar.toolbar);
        setTitle(R.string.app_name);

        sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);

        // Display area
        binding.gpsButton.setOnClickListener(this::openLocationSettings);

        // Button area
        binding.shareButton.setOnClickListener(this::shareLocation);
        binding.copyButton.setOnClickListener(this::copyLocation);
        binding.viewButton.setOnClickListener(this::viewLocation);

        // Set default values for preferences
        PreferenceManager.setDefaultValues(this, R.xml.preferences, false);

        locManager = (LocationManager) getSystemService(LOCATION_SERVICE);
    }

    @Override
    protected void onStop() {
        super.onStop();
        try {
            locManager.removeUpdates(locListener);
        } catch (SecurityException e) {
            Log.e(TAG, "Failed to stop listening for location updates", e);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        startRequestingLocation();
        updateLocation();
    }

    private final ActivityResultLauncher<String> requestPermissionLauncher =
            registerForActivityResult(new RequestPermission(), isGranted -> {
                if (isGranted) {
                    startRequestingLocation();
                } else {
                    Toast.makeText(this, R.string.permission_denied, Toast.LENGTH_SHORT).show();
                    finish();
                }
            });

    // ----------------------------------------------------
    // UI
    // ----------------------------------------------------
    private void updateLocation() {
        // Trigger a UI update without changing the location
        updateLocation(lastLocation);
    }

    private void updateLocation(Location location) {
        boolean locationEnabled = locManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
        boolean waitingForLocation = locationEnabled && !validLocation(location);
        boolean haveLocation = locationEnabled && !waitingForLocation;

        // Update display area
        binding.gpsButton.setVisibility(locationEnabled ? View.GONE : View.VISIBLE);
        binding.progressTitle.setVisibility(waitingForLocation ? View.VISIBLE : View.GONE);
        binding.progressBar.setVisibility(waitingForLocation ? View.VISIBLE : View.GONE);
        binding.detailsText.setVisibility(haveLocation ? View.VISIBLE : View.GONE);

        // Update buttons
        binding.shareButton.setEnabled(haveLocation);
        binding.copyButton.setEnabled(haveLocation);
        binding.viewButton.setEnabled(haveLocation);

        if (haveLocation) {
            String newline = System.getProperty("line.separator");
            binding.detailsText.setText(String.format("%s: %s%s%s: %s (%s)%s%s: %s (%s)",
                    getString(R.string.accuracy), getAccuracy(location), newline,
                    getString(R.string.latitude), getLatitude(location), getDMSLatitude(location), newline,
                    getString(R.string.longitude), getLongitude(location), getDMSLongitude(location)));

            lastLocation = location;
        }
    }

    // ----------------------------------------------------
    // DialogInterface Listeners
    // ----------------------------------------------------
    private class onClickShareListener implements OnClickListener {
        @Override
        public void onClick(DialogInterface dialog, int i) {
            shareLocationText(formatLocation(lastLocation, getResources().getStringArray(R.array.link_options)[i]));
        }
    }

    private class onClickCopyListener implements OnClickListener {
        @Override
        public void onClick(DialogInterface dialog, int i) {
            copyLocationText(formatLocation(lastLocation, getResources().getStringArray(R.array.link_options)[i]));
        }
    }

    //-----------------------------------------------------
    // Menu related methods
    //-----------------------------------------------------

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        this.getMenuInflater().inflate(R.menu.menu_toolbar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_settings) {
            Intent intentSettingsActivity = new Intent(this, SettingsActivity.class);
            this.startActivity(intentSettingsActivity);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    // ----------------------------------------------------
    // Actions
    // ----------------------------------------------------
    public void shareLocation(View view) {
        if (!validLocation(lastLocation)) {
            return;
        }

        String linkChoice = sharedPrefs.getString("prefLinkType", "");

        if (linkChoice.equals(getResources().getString(R.string.always_ask))) {
            new Builder(this).setTitle(R.string.choose_link)
                    .setCancelable(true)
                    .setItems(R.array.link_names, new onClickShareListener())
                    .create()
                    .show();
        } else {
            shareLocationText(formatLocation(lastLocation, linkChoice));
        }
    }

    public void copyLocation(View view) {
        if (!validLocation(lastLocation)) {
            return;
        }

        String linkChoice = sharedPrefs.getString("prefLinkType", "");

        if (linkChoice.equals(getResources().getString(R.string.always_ask))) {
            new Builder(this).setTitle(R.string.choose_link)
                    .setCancelable(true)
                    .setItems(R.array.link_names, new onClickCopyListener())
                    .create()
                    .show();
        } else {
            copyLocationText(formatLocation(lastLocation, linkChoice));
        }
    }

    public void viewLocation(View view) {
        if (!validLocation(lastLocation)) {
            return;
        }

        String uri = formatLocation(lastLocation, "geo:{0},{1}?q={0},{1}");

        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
        startActivity(Intent.createChooser(intent, getString(R.string.view_location_via)));
    }

    public void openLocationSettings(View view) {
        if (!locManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            startActivity(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS));
        }
    }

    // ----------------------------------------------------
    // Helper functions
    // ----------------------------------------------------
    public void shareLocationText(String string) {
        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_SEND);
        intent.putExtra(Intent.EXTRA_TEXT, string);
        intent.setType("text/plain");
        startActivity(Intent.createChooser(intent, getString(R.string.share_location_via)));
    }

    public void copyLocationText(String string) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        if (clipboard != null) {
            ClipData clip = ClipData.newPlainText(getString(R.string.app_name), string);
            clipboard.setPrimaryClip(clip);
            Toast.makeText(getApplicationContext(), R.string.copied, Toast.LENGTH_SHORT).show();
        } else {
            Log.e(TAG, "Failed to get the clipboard service");
            Toast.makeText(getApplicationContext(), R.string.clipboard_error, Toast.LENGTH_SHORT).show();
        }
    }

    private void startRequestingLocation() {
        if (!locManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            return;
        }

        final String permission = Manifest.permission.ACCESS_FINE_LOCATION;
        if (ActivityCompat.checkSelfPermission(this, permission)
                != PackageManager.PERMISSION_GRANTED) {
            requestPermissionLauncher.launch(permission);
            return;
        }

        // GPS enabled and have permission - start requesting location updates
        locManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, locListener);
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    private boolean validLocation(Location location) {
        if (location == null) {
            return false;
        }

        // Location must be from less than 30 seconds ago to be considered valid
        if (Build.VERSION.SDK_INT < 17) {
            return System.currentTimeMillis() - location.getTime() < 30e3;
        } else {
            return SystemClock.elapsedRealtimeNanos() - location.getElapsedRealtimeNanos() < 30e9;
        }
    }

    private String getAccuracy(Location location) {
        float accuracy = location.getAccuracy();
        if (accuracy < 0.01) {
            return "?";
        } else if (accuracy > 99) {
            return "99+";
        } else {
            return String.format(Locale.US, "%2.0fm", accuracy);
        }
    }

    private String getLatitude(Location location) {
        return String.format(Locale.US, "%2.5f", location.getLatitude());
    }

    private String getDMSLatitude(Location location) {
        double val = location.getLatitude();
        return String.format(Locale.US, "%.0f° %2.0f′ %2.3f″ %s",
                Math.floor(Math.abs(val)),
                Math.floor(Math.abs(val * 60) % 60),
                (Math.abs(val) * 3600) % 60,
                val > 0 ? "N" : "S"
        );
    }

    private String getDMSLongitude(Location location) {
        double val = location.getLongitude();
        return String.format(Locale.US, "%.0f° %2.0f′ %2.3f″ %s",
                Math.floor(Math.abs(val)),
                Math.floor(Math.abs(val * 60) % 60),
                (Math.abs(val) * 3600) % 60,
                val > 0 ? "E" : "W"
        );
    }

    private String getLongitude(Location location) {
        return String.format(Locale.US, "%3.5f", location.getLongitude());
    }

    private String formatLocation(Location location, String format) {
        return MessageFormat.format(format,
                getLatitude(location), getLongitude(location));
    }
}
